In [26]:
###Import all the required libraries
import pandas as pd
from pandas import json_normalize
import json
from datetime import datetime

json_file_path = 'C:/Users/princ/PycharmProjects/pythonProject/Assessment/users.json'

# Read JSON file line by line into a list
with open(json_file_path, 'r') as file:
    lines = file.readlines()

# Load each line as a separate JSON object
data = [json.loads(line) for line in lines]

# json_normalize to flatten the nested structure
users = json_normalize(data)
users.head(10)

active      role signUpSource state                  _id.$oid  \
0    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3    True  consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
5    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
6    True  consumer        Email    WI  5ff1e1e8cfcf6c399c274ad9   
7    True  consumer        Email    WI  5ff1e1b7cfcf6c399c274a5a   
8    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
9    True  consumer        Email    WI  5ff1e1f1cfcf6c399c274b0b   

   createdDate.$date  lastLogin.$date  
0      1609687444800     1.609688e+12  
1      1609687444800     1.609688e+12  
2      1609687444800     1.609688e+12  
3      1609687530554     1.609688e+12  
4      1609687444800     1.609688e+12  
5      1609687444800     1.609688e+12  
6      1609687528354     1.609688e+12  
7      1609687479626     1.609687e+12  
8      1609687444800     1.609688e+12  
9      1609687537564     1.609688e+12

In [27]:
users.rename(columns={
    '_id.$oid': 'user_id',
    'createdDate.$date':'created_date',
    'lastLogin.$date':'lastlogin_data'}, inplace=True)
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   active          495 non-null    bool   
 1   role            495 non-null    object 
 2   signUpSource    447 non-null    object 
 3   state           439 non-null    object 
 4   user_id         495 non-null    object 
 5   created_date    495 non-null    int64  
 6   lastlogin_data  433 non-null    float64
dtypes: bool(1), float64(1), int64(1), object(4)
memory usage: 23.8+ KB


In [28]:
users['created_date'] = users['created_date'].dropna().apply(lambda x: datetime.fromtimestamp(x / 1000.0))
users['lastlogin_data'] = users['lastlogin_data'].dropna().apply(lambda x: datetime.fromtimestamp(x / 1000.0))

In [29]:
excel_file_path = 'C:/Users/princ/PycharmProjects/pythonProject/Assessment/users.xlsx'
users.to_excel(excel_file_path, index=False)
users.head(10)

active      role signUpSource state                   user_id  \
0    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3    True  consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
5    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
6    True  consumer        Email    WI  5ff1e1e8cfcf6c399c274ad9   
7    True  consumer        Email    WI  5ff1e1b7cfcf6c399c274a5a   
8    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
9    True  consumer        Email    WI  5ff1e1f1cfcf6c399c274b0b   

             created_date          lastlogin_data  
0 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
1 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
2 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
3 2021-01-03 10:25:30.554 2021-01-03 10:25:30.597  
4 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
5 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
6 2021-01-03 10:25:28.354 2021-01-03 10:25:28.392  
7 2021-01-03 10:24:39.626 2021-01-03 10:24:39.665  
8 2021-01-03 10:24:04.800 2021-01-03 10:25:37.858  
9 2021-01-03 10:25:37.564 2021-01-03 10:25:37.599

In [23]:
users.isna().sum()

active             0
role               0
signUpSource      48
state             56
receipt_id         0
created_date       0
lastlogin_data    62
dtype: int64

In [30]:
percentage_null_values= users.isnull().mean()
print("Percentage of null values:")
for key, value in percentage_null_values.items():
  if value>0:
    print(f"{key}: {round(value * 100, 2)}%")

Percentage of null values:
signUpSource: 9.7%
state: 11.31%
lastlogin_data: 12.53%


SignUpSource: Null values in this column might affect the evalution of marketing strategies and user segmentation.

In [34]:
"""Checking duplicates"""
duplicates = users[users.duplicated()]
if not duplicates.empty:
    print('Duplicate rows (except for their first occurrence):')
    print(duplicates)
else:
    print('No duplicate rows found.')

Duplicate rows (except for their first occurrence):
     active         role signUpSource state                   user_id  \
1      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
4      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
5      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
8      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
..      ...          ...          ...   ...                       ...   
490    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
491    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
492    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
493    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
494    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   

               created_date          lastlogin_data  
1   2021-01-03 10

In [33]:
print(users["user_id"].nunique())
print(users["signUpSource"].unique())
print(users["state"].unique())
print(users['signUpSource'].value_counts())
print(users['state'].value_counts())

212
['Email' 'Google' nan]
['WI' 'KY' 'AL' 'CO' 'IL' nan 'OH' 'SC' 'NH']
signUpSource
Email     443
Google      4
Name: count, dtype: int64
state
WI    396
NH     20
AL     12
OH      5
IL      3
KY      1
CO      1
SC      1
Name: count, dtype: int64


In [40]:
"""Checking duplicates"""
duplicates = users[users.duplicated()]
if not duplicates.empty:
    print('Duplicate rows (except for their first occurrence):')
    print(duplicates.value_counts())
    print(f"No of duplicate user_id's: {duplicates['user_id'].count()}")
else:
    print('No duplicate rows found.')

Duplicate rows (except for their first occurrence):
active  role         signUpSource  state  user_id                   created_date             lastlogin_data         
True    fetch-staff  Email         NH     5fc961c3b8cfca11a077dd33  2020-12-03 17:08:03.936  2021-02-26 17:39:16.799    19
        consumer     Email         WI     5ff5d15aeb7c7d12096d91a2  2021-01-06 10:03:54.680  2021-01-06 10:08:10.009    17
                                          600fb1ac73c60b12049027bb  2021-01-26 01:07:40.879  2021-01-26 01:11:23.950    15
                                          5ff1e194b6a9d73a3a9f1052  2021-01-03 10:24:04.800  2021-01-03 10:25:37.858    10
                                          600987d77d983a11f63cfa92  2021-01-21 08:55:35.327  2021-01-21 08:59:21.063     8
                                          600056a3f7e5b011fce897b0  2021-01-14 09:35:15.207  2021-01-14 09:37:10.767     7
                                          60189c74c8b50e11d8454eff  2021-02-01 19:27:32.642  

As per this data, majority of the users that have signed up reside in Wisconsin. Data quality issues found:
1. More than half of the user records(out of the total 495) are redundant which might result in skewed analysis and inefficient marketing efforts.
2. Another, less pressing issue: small percentage of missing values in signUpSource and state columns